qintelqwatch_connector.py

```diff
@@ -0,0 +1,412 @@
+
+        # initial poll - start time
+        if not start_time:
+            if not (1 <= self.qwatch_initial_window <= 90):
+                raise Exception(
+                    'Config Error: QWatch Initial Ingestion Window '
+                    'parameter must be between 1 and 90'
+                )
+
+            start_time = datetime.utcnow() - \
+                         timedelta(days=self.qwatch_initial_window)
+            start_time = int(start_time.timestamp())
+
+        end_time = params.get('end_time')
+        if not end_time:
+            end_time = int(datetime.utcnow().timestamp())
+        else:
+            end_time = int(end_time / 1000)  # convert from ms
+
+        return start_time, end_time
+
+    def _make_qwatch_params(self, param):
+        start_time, end_time = self._make_qwatch_times(param)
+
+        req_params = {
+            'date[start]': start_time,
+            'date[end]': end_time,
+            'meta[total]': True,
+            'stats': True
+        }
+
+        if param.get('artifact_count'):
+            limit = int(param['artifact_count'])
+            if limit < 50:
+                self.save_progress('Adjusting artifact limit to minimum of 50')
+                limit = 50
+
+            if limit > 10000:
+                self.save_progress(
+                    'Adjusting artifact limit to maximum of 10000')
+                limit = 1000
+
+            req_params['limit'] = limit
+
+        return req_params
+
+    def _process_timestamps(self, exposure):
+
+        ret_ts = {}
+
+        # start time
+        timestamps = exposure['attributes']['timestamps']
+        for ts in timestamps:
+            if ts['context'] == 'loaded':
+                ret_ts['loaded'] = f'{ts["iso"]}Z'
+
+        # first seen - cred
+        timestamps = exposure['meta']['stats']['credential']['timestamps']
+        for ts in timestamps:
+            if ts['context'] == 'first_seen':
+                ret_ts['first_seen'] = f'{ts["iso"]}Z'
+
+        return ret_ts
+
+    def _process_qwatch_exposure(self, exposure):
+
+        return_data = {}
+
+        timestamps = self._process_timestamps(exposure)
+
+        return_data['credential'] = exposure['attributes']['login_name']
+
+        return_data['password'] = None
+        if self.plaintext_passwords:
+            return_data['password'] = exposure['attributes']['password']
+
+        return_data['source_name'] = exposure['attributes']['source_name']
+        return_data['last_seen'] = timestamps.get('loaded')
+        return_data['first_seen'] = timestamps.get('first_seen')
+
+        return return_data
+
+    def _set_severity(self, timestamps, cef):
+
+        source_name = cef['exposureSourceName']
+        if source_name.lower().startswith('malware'):
+            return 'high'
+
+        if timestamps['loaded'] == timestamps['first_seen']:
+            return 'high'
+
+        return 'medium'
+
+    def _process_exposures_ingest(self, data):
+
```

Comment:
This method is quite long. Would it be possible to add docstring here?

Reply:
Done